### DISCUSSION TITLE
Is it possible to automatically create a processor module in Ghidra by using a processor specification sheet?

### DISCUSSION THREAD

**Discussion by Jfriar10** [NationalSecurityAgency/ghidra#5814] [Q&A]
At the start of the software reverse engineering process, engineers must use Ghidra processor modules. However, if the processor module is unavailable, engineers must manually create it from scratch, a tedious and time-consuming process that can take months to complete. Due to the labor-intensive nature of this process, errors can occur, and it becomes difficult to check for bugs within the module. 

Is there any way to automatically create a processor module in Ghidra using a processor specification sheet?

Any help would be greatly appreciated. 
Thank you!

**Comment by Tim---** [comment]
Hi,

If the only resource you have is the instruction set architecture in PDF format, you're not going to automate a lot.
When copy-pasting a line often ends up with pasting a bunch of other characters, this format is as useful as it looks.
AFAIK, the instruction sets are too different between several manufacturers, so noone tried to do a generic parsing or OCR on the ISA files.

The most "parsable" input that you can have is probably the source of binutils/gcc if your processor is supported.
I think I saw some projects take this as input to do the heavy-lifting of disassembly.

I tried this a bit too and in my experience it was finally faster to do the whole process manually. The only important requirement is to have enough screen space to have both the ISA and pspec files side by side.

For a standard RISC processor without an exotic design, you can get a decent result in a week. It's good enough that you can analyze functions, even though you have to do some back and forth to fix a typo or implement one instruction.

Also the most important part of avoiding errors is implementing the pcodetests if you happen to have a compiler for your architecture. This will catch a lot of typos and bugs, and give you an indication of the maturity of your implementation.

**Reply by Jfriar10** [reply]
This is very helpful, thank you for your response, Tim!
